tests/optional_scalars2_generated.rs

```diff
@@ -18,69 +18,74 @@ pub mod optional_scalars {
   extern crate flatbuffers;
   use self::flatbuffers::EndianScalar;
 
-#[allow(non_camel_case_types)]
-#[repr(i8)]
-#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
-pub enum OptionalByte {
-  None = 0,
-  One = 1,
-  Two = 2,
-
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct OptionalByte(pub i8);
```

Comment:
Please use proper enums for enums.

Reply:
> Please use proper enums for enums.

Hard to do due to endianness. There's no difference other than exhaustive pattern match between `#[repr(u*/i*)]` enums and `Struct(u*/i*)` with a bunch of constants to emulate enum. But the struct could be safely read from raw bytes and byteswapped. There are other way to work around this, but this is simply the easiest and safest. Again, the only thing you lose is exhaustiveness in matching against such enum, but all those enums are anyway expected to be inexhaustive, because schema evolution go brr.

Reply:
I'm in favor of `TryFrom<underlying type>` with exhaustive enum, rather than this. We loose some performance but we gain safety and proper enum representation.

Reply:
> I'm in favor of `TryFrom<underlying type>` with exhaustive enum, rather than this. We loose some performance but we gain safety and proper enum representation.

Struct are always little-endian. How would you store the enum in such a struct?

Reply:
FB does not store enums but underlying representation. The getter is what transforms the bytes to enum, hence the getter calling try_into() on the buffer will do the trick in a portable fasion.

Reply:
> FB does not store enums but underlying representation. The getter is what transforms the bytes to enum, hence the getter calling try_into() on the buffer will do the trick in a portable fasion.

What about ergonomics? Struct could have been created from trusted data, parsed from untrusted data or created by the user. In the latter case getter is not needed, and in the first two cases I'd like the getter to be infallible.

What about performance? If the data is trusted, I'd like to not check enum values. If the data is untrusted, I'd like to check enum only once instead of every time I get it from the struct. Do you expect users to write something like this?

```rust
struct ParsedStruct {
    a: Enum,
}

impl ParsedStruct {
    fn try_from(unparsed: UnparsedStruct) -> Option<Self> {
        Enum::try_from(unparsed.enum.to_le()).ok().map(|en| Self { a: en })
    }

    unsafe fn from_unchecked(unparsed: UnparsedStruct) -> Self {
        Self { a: unsafe { transmute(unparsed.enum.to_le()) } }
    }
}
```

It is inconvenient. All this parsing should be generated, not hand-written by user. But again, this is *not* a PR about rewriting `Follow`.

Reply:
I don't know why you wish to use parsed structs, since it goes against the idea of zero parsing of FB. Instead, use fallible getters with Results, along with infallible. How to do the latter is an open question. panic seems the reasonable approach here, if we don't want to use unsafe.

Another option is a validator, which will return a safe version of the buffer, where every getter is infallible and safe.

Reply:
> proper enum representation

Those enums are expected to be inexhaustive for schema evolution reasons. There is essentially no difference between inexhaustive enum and `u*/i*` with constants to emulate enum.

Reply:
> why you wish to use parsed structs

Because using parsed structs is fast, cool and much simpler than using unparsed structs that may fail on every getter. Plus, getters of items in the struct are totally safe and panic-free. Parsing struct may fail, but using struct would never fail.

> Another option is a validator

Yes. It is even better if it is parser, because using parser requires zero unsafe usage, unlike using validator + getter pair.

Reply:
Well, parsing in FB is the responsibility of the user. FB provides typed getters into the buffer. That's the thing which needs to be solved somehow in Rust. If the results then gets put into a struct or not - that's no longer FB concern. But going back to the topic of enums - I've already commented earlier that to have both forwards-compatible enums and idiomatic Rust API, we need a fallible getter.

Reply:
> Well, parsing in FB is the responsibility of the user. FB provides typed getters into the buffer.

Quoting myself: "nothing tough, you just have to rewrite `Follow`". Really, stop that, just look at how `Follow` is done [here](https://github.com/ImmemorConsultrixContrarie/cool_flatbuffers). It is not very hard to generate
```rust
if Q::follow(buf, loc + q_offset).is_some() && W::follow(buf, loc + w_offset).is_some() && … {
    Some(unsafe { Self::follow_unchecked(buf, loc) })
} else {
    None
}
```

I'm saying it one last time: this is **not** a PR about rewriting `Follow`.

I won't talk about fallible field getters anymore, because failing to get a field from the struct or writing `unsafe` to get field *sucks*. All functions would be tainted by `Result<_, ()>`, by unsafe, or by `unwrap`s. Structs should be parsed once, not validated and getted by `unsafe` or `unwrap` later, period. It is safer, because you may forget to validate the data, but you can't forget to parse it; it also leads to less programmer-written code and less machine code.

Reply:
In other words, you would like something akin to the previous idea of a validator returning a safe buffer.

Reply:
> In other words, you would like something akin to the previous idea of a validator returning a safe buffer.

Not buffer, struct. With safe infallible getters.

Reply:
That will most likely never happen, because of goes against one of the design principles of FB - avoiding parsing. This principle is upheld across all supported languages and Rust should not be an exception. That's why there are several discussions already about making a better Rust API.

Reply:
> avoiding parsing

Lolwut? There's no parsing, because making a generic parser in C would be hard. In Rust we don't have those problems.

It is also not going against the "fastest zero-copy" way, because there's that `follow_unchecked`, and if you trust the data, you can skip parsing checks.

Reply:
In such case, I suggest reading https://google.github.io/flatbuffers/ first to see what the philosophy behind FB is. Again - this is how it works across every supported language. If you want to a parsed representation, you are free to write a parser for your needs. Right now, this PR tries to address one of the symptoms of a general problem of Rust FB. Let's not go offtopic about parsed vs unparsed data, since this is not the problem here.

Reply:
> In such case, I suggest reading https://google.github.io/flatbuffers/ first to see what the philosophy behind FB is. Again - this is how it works across every supported language. If you want to a parsed representation, you are free to write a parser for your needs. Right now, this PR tries to address one of the symptoms of a general problem of Rust FB. Let's not go offtopic about parsed vs unparsed data, since this is not the problem here.

Funny thing: you started this, saying: "it must be `Result`, so we can parse the enum, not just get it, even if it is the data we trust". Now again, who's going against the philosophy?

That "no-parsing" is about trusted data, or the validator would not exist at all. In Rust we simply can make generic parser. In C we can't, thus there's that validator.

And the problem here is that we have data and we want get it in a cool infallible way. Without UB. Without panics. Without writing "unsafe" each time we get it. Because, well, it is the data we trust/validated/parsed. Why do you want to have a `Result` here?

Reply:
The problem here is that a validator is needed for languages which can exhibit UB. You have one for C++, so you can have a zero-parsing approach. You don't have one in Rust, so you can't have a zero-parsing approach. That's why there are some ideas floating around, one of them being a Result, other creating a validator, etc. Until we have a validator, we cannot trust a buffer. Since we cannot trust a buffer, we need fallible getters or we risk panics or unsoudness. If users want to put that data in a parsed struct - that's their choice to make. FB is only responsible for serialization (which btw. is somewhat problematic in languages like Java).

Reply:
> You don't have one in Rust

A little bit of FFI magic, and you have validator. Validator absence is not an argument here.

> need fallible getters

Why? We can have a `Wrapper(u8)`, that works exactly like inexhaustive enum works, but does not need a fallible getter. Why do you want to make it fallible so much?

Reply:
FFI is not a solution - a pure Rust validator is. If you wish to contribute, you can create one. Also, don't be focused solely on enums. We have other data types which need to be properly interpreted from the underlying buffer, which includes handling all the error cases (invalid buffer size, invalid data, range errors, etc.).

Reply:
> don't be focused solely on enums

*Here in this PR* we are talking about enums. Nothing but enums. Now, please, tell me: why do you want a fallible enum getter instead of infallible?

Reply:
Like I said before: this PR tries to address one of the symptoms of a general problem of Rust FB. We should address the whole problem, not just a part of it. And I'm for fallible getters as a temporary solution until we have a proper validator and trusted buffers. If the underlying data contains a value which cannot be represented by an enum value in Rust, we have an unsound api. How to deal with that particular situation depends on a language. In some languages, we don't really care and cast one to another. In some, native enums are avoided. In Rust both of these approaches are not idiomatic. We should have a solution which solves this issue and can be extrapolated to all types, otherwise we get a mess of different approaches.

Reply:
> Like I said before: this PR tries to address one of the symptoms of a general problem of Rust FB. We should address the whole problem, not just a part of it. And I'm for fallible getters as a temporary solution until we have a proper validator and trusted buffers. If the underlying data contains a value which cannot be represented by an enum value in Rust, we have an unsound api. How to deal with that particular situation depends on a language. In some languages, we don't really care and cast one to another. In some, native enums are avoided. In Rust both of these approaches are not idiomatic. We should have a solution which solves this issue and can be extrapolated to all types, otherwise we get a mess of different approaches.

So, you wanna start an issue about the sad state or current Rust lib? Why do you want to make it inside the PR? Do you think, that PR named "rework enums in rust" would be extended to change the whole lib instead of just enums?

And you did not answer the question. What are bonuses of using native inexhaustive Rust enums? I cannot see any difference in matching.

If you don't have any arguments to support your statement "fallible **enum** getter is better than infallible", could you, please, mark this conversation as solved?

Reply:
I don't want to start the conversation about current Rust implementation, because it has been already started in other places. My argument for a fallible getter is that:

- You'll get enum representation as a real enum type, instead of a generic integer. Java implementation is an example where constants are used and it's quite awkward to use numbers when a language has enum type support. W can finally avoid wonderful constructs like `int result = Color.Red + Direction.South;`.
- For cases when we want to be sure the buffer only contains the values we support, Result will give the natural answer.
- On the other hand, if we don't care about other values, we can simply ignore the error.
- Using a Result gives all the benefits of method chaining etc.

Reply:
> generic integer

It's a newtype wrapper. In Rust you can't write `Color(1) + Radius(1)` without explicitly writing `impl Add<Radius> for Color`. Can't see a problem here.

> For cases when we want to be sure the buffer only contains the values we support, Result will give the natural answer.

And what do we do, if we trust the data and don't care? Generate more code with `unsafe` or `unwrap`? You are just trying to put a dead body into the hole instead of repairing the ship. If you do so, please, put bodies in all holes under the waterline: write PR and make the whole `Follow` trait fallible. And make everything that can fail fallible too. Do not start this from insisting on putting a body into the hole which is not under waterline and does not actually leak. This one could be fixed later.

> On the other hand, if we don't care about other values, we can simply ignore the error.

How? You have `Result`. You can't ignore it. You must either match or unwrap. Or make some dirty things like `transmute`. Or use some sugar like `?`. You can't just ignore the error.

> Using a Result gives all the benefits of method chaining etc.

Using `T` instead of `Result<T, ()>` makes you write less code and generate less machine code. What is that "method chaining" and how would it help, if you need `T`, but get `Result<T>`? You need to work with `T`, not with `Result`.

Reply:
If we are sure the buffer is correct, e.g. by validating it, which is not possible now, we can create infallible getters. That's the end goal. As for ignoring the error, yes - you must use match. Just like a every fallible operation. If you don't want the burden, you are free to use unsafe or another language. Nevertheless, we both stated our opinions.

Reply:
> If we are sure the buffer is correct, e.g. by validating it

What if we are sure it is correct, because we created it?

Reply:
Nothing prevents you from using an unsafe getter then, since you know it's safe.

Reply:
> Nothing prevents you from using an unsafe getter then, since you know it's safe.

And now we are talking about rewriting `Follow` again. Why? Let's compare those two pieces of code:

rewritten `Follow`

```rust
let root = unsafe { get_root_unchecked::<Root>(buffer) };
let x: i32 = root.x();
println!("{}", x);
match root.color() {
    Color::Red => (),
    Color::Blue => panic!("Ugh, blue!"),
    _ => (),
}
```

your "fallible or unsafe"

```rust
let root = unsafe { get_root_unchecked::<Root>(buffer) };
let x: i32 = unsafe { root.x_unchecked() };
println!("{}", x);
match unsafe { root.color_unchecked() } {
    Color::Red => (),
    Color::Blue => panic!("Ugh, blue!"),
    _ => (),
}
```

I would prefer to write the first piece. Really. The same goes for validating the data: I'd prefer `get_root(buffer)?` to be the only place where I should write `?`. And rewriting `Follow` would take the same time as making everything that should be fallible — fallible.

Or, maybe you'd prefer the second version? Especially when reference to `root` goes from function to function and you don't know whether it is sound or unsound to get something from the `root` without checks?

Reply:
My preferable version (assuming no validation) would contain two getters:

```
fn get_enum() -> Result<Enum, Error>;
unsafe fn get_enum_unchecked() -> Enum;
```

Then you have the option to call either two:

```
let e1 = root.get_enum()?;
let e2 = unsafe { root.get_enum_unchecked() };
```

Reply:
> two getters

Wait. Are you saying, that instead of parsing (or one call to unsafe for trusted data), i.e.
```rust
fn ytr(s: Option<&[u8]>) {
    if let Some(s) = s {
        asd(s, s.len());
    }
}

/// # Safety
///
/// UB if `s` is `None`.
unsafe fn qwe(s: Option<&[u8]>) {
        let s: &[u8] = unsafe { transmute(s) };
        asd(s, s.len());
    }
}

fn asd(s: &[u8], len: usize) {
    println!("len: {}; sloice: {}", len. s);
}
```

you would prefer valudation (either fallible or unsafe), i.e.
```rust
fn ytr(s: Option<&[u8]>) {
    if s.is_none() { return; }
    let ss = s.unwrap();
    hgf(s, ss.len());
}

/// # Panic
///
/// Panics if `s` is `None`.
fn hgf(s: Option<&[u8]>) {
    println!("len: {}; sloice: {}", len. s.unwrap());
}

/// # Safety
///
/// UB if `s` is `None`.
unsafe fn qwe(s: Option<&[u8]>) {
    let ss: &[u8] = unsafe { transmute(s) };
    unsafe { asd(s, ss.len()); }
}
/// # Safety
///
/// UB if `s` is `None`.
unsafe fn asd(s: Option<&[u8]>, len: usize) {
    let s: &[u8] = unsafe { transmute(s) };
    println!("len: {}; sloice: {}", len. s);
}
```

?

Reply:
Please use my proposed functions to ask your question, instead of a contrived example.

Reply:
> Please use my proposed functions to ask your question, instead of a contrived example.

I'm asking just that: do you prefer polluting `Result`, `Option`, or `unsafe` everywhere (literally in each function that works with flatbuffer data, and that could be a surprisingly big number of functions in some projects), or do you prefer keeping it in one place? Because that's the difference between making getters for all fields for each struct and table (including nested structs and table) fallible and making only top-struct getter fallible.

Reply:
Keeping it in one place involves having the buffer validated, which was already explained. Let's not go in circles. We need either that or a safe way to get properly typed data from unknown buffers.

Reply:
> Keeping it in one place involves having the buffer validated, which was already explained. Let's not go in circles. We need either that or a safe way to get properly typed data from unknown buffers.

The time we need to "rework `Follow`" and the time we need to "make all field getters dual fallible + unsafe" are equal. The only win point of the second approach is that it could be split into many little PRs. But do you really want to get the result of second approach instead of one dual `fallible + unsafe` getter for top-level struct?

Reply:
Again, please read what was said before - I **do** prefer one fallible method which will return a validated buffer/struct, which in turn would have infallible getters. The thing I do not agree with you is claiming it will take the same time. And I do this based on the Rust verifier supplied by cfb and the amount of time they took to make it. But, if someone choses to create something returning this verified buffer/struct, I'm all for it. If you really can do it in such short amount of time - by all means, create a PR and everybody wins.

Reply:
Oh, well, okay. The codegen is the problem though. It is not very hard to do, except that I'm awful at cpp, thus, well, it is very hard to do. And while this PR is not ready, I'd really like to see that "make fallible + unsafe getters for everything" as an issue instead of stopper for the PR that would close a big soundness hole.

Reply:
Well, the soundness hole with enums now can be solved by safe fallible + unsafe infallible getters with proper enum types, or like the way it's proposed now, disregarding the enum type completely. My opinion is that we should use the real enum type for enums and Result for fallible conversion. We also know your opinion. Let's see what others have to say and decide then.

Reply:
Resolving this conversation as per the conversation below